File: benchmarks/binarytrees/csharp/BinaryTrees.cs
using System;
using System.Threading.Tasks;

class BinaryTrees
{
    class TreeNode
    {
        readonly TreeNode left, right;

        internal TreeNode(int d)
        {
            if(d != 0)
            {
                left = new TreeNode(d - 1);
                right = new TreeNode(d - 1);
            }
        }

        internal static int Check(TreeNode current)
        {
            int c = 0;
            while (current != null)
            {
                c += Check(current.right) + 1;
                current = current.left;
            }
            return c;
        }
    }

    const int MinDepth = 4;
    const int NoTasks = 4;

    public static void Main(string[] args)
    {
        int maxDepth = args.Length == 0 ? 10
            : Math.Max(MinDepth + 2, int.Parse(args[0]));


        long startTime = DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond;

        Console.WriteLine(string.Concat("stretch tree of depth ", maxDepth + 1,
            "\t check: ", TreeNode.Check(new TreeNode(maxDepth + 1))));

        var longLivedTree = new TreeNode(maxDepth);

        var results = new string[(maxDepth - MinDepth) / 2 + 1];

        for (int i = 0; i < results.Length; i++)
        {
            int depth = i * 2 + MinDepth;
            int n = (1 << maxDepth - depth + MinDepth) / NoTasks;
            var tasks = new Task<int>[NoTasks];
            for (int t = 0; t < tasks.Length; t++)
            {
                tasks[t] = Task.Run(() =>
                {
                    var check = 0;
                    for (int i = n; i > 0; i--)
                        check += TreeNode.Check(new TreeNode(depth));
                    return check;
                });
            }
            var check = tasks[0].Result;
            for (int t = 1; t < tasks.Length; t++)
                check += tasks[t].Result;
            results[i] = string.Concat(n * NoTasks, "\t trees of depth ",
                depth, "\t check: ", check);
        }

        for (int i = 0; i < results.Length; i++)
            Console.WriteLine(results[i]);

        Console.WriteLine(string.Concat("long lived tree of depth ", maxDepth,
            "\t check: ", TreeNode.Check(longLivedTree)));

        long end = DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond;

        Console.WriteLine(string.Concat("time: ", end - startTime, "ms"));
    }
}

File: benchmarks/binarytrees/rust_backdrop/binarytrees.rs

use backdrop_arc::TrashThreadStrategy;
use backdrop_arc::Arc;
use snmalloc_rs::SnMalloc;

#[global_allocator]
static ALLOC: SnMalloc = SnMalloc;

#[allow(dead_code)]
pub struct Node {
    item: i64,
    left: Option<Arc<Node, TrashThreadStrategy>>,
    right: Option<Arc<Node, TrashThreadStrategy>>,
}

impl Node {
    fn check_tree(&self) -> usize {
        if self.left.is_none() {
            return 1;
        }

        1 + self.left.as_ref().unwrap().check_tree() + self.right.as_ref().unwrap().check_tree()
    }
}

fn create_tree(depth: i64) -> Arc<Node, TrashThreadStrategy> {
    if 0 < depth {
        let node = Arc::new(Node {
            item: 0,
            left: Some(create_tree(depth - 1)),
            right: Some(create_tree(depth - 1)),
        });

        node
    } else {
        let node = Node {
            item: 0,
            left: None,
            right: None,
        };

        Arc::new(node)
    }
}

fn bench_parallel() {
    let mut n = 0;
    if let Some(arg) = std::env::args().skip(1).next() {
        if let Ok(x) = arg.parse::<usize>() {
            n = x;
        }
    }

    let min_depth = 4;
    let max_depth = if n < (min_depth + 2) {
        min_depth + 2
    } else {
        n
    };

    let start = std::time::Instant::now();
    let stretch_depth = max_depth + 1;

    {
        println!(
            "stretch tree of depth {}\t check: {}",
            stretch_depth,
            create_tree(stretch_depth as _).check_tree()
        );
    }

    let long_lasting_tree = create_tree(max_depth as _);
    use parking_lot::Mutex;
    let results: Arc<_, TrashThreadStrategy> = Arc::new(
        (0..(max_depth - min_depth) / 2 + 1)
            .map(|_| Mutex::new(String::new()))
            .collect::<Vec<_>>(),
    );

    std::thread::scope(|scope| {
        let mut d = min_depth;

        while d <= max_depth {
            let depth = d;
            let cloned = Arc::clone(&results);
            scope.spawn(move || {
                let iterations = 1 << (max_depth - depth + min_depth);
                let mut check = 0;
                for _ in 1..=iterations {
                    let tree_node = create_tree(depth as _);
                    check += tree_node.check_tree();
                }

                *cloned[(depth - min_depth) / 2].lock() = format!(
                    "{}\t trees of depth {}\t check: {}",
                    iterations, depth, check
                );
            });

            d += 2;
        }
    });

    for result in results.iter() {
        println!("{}", *result.lock());
    }
    println!(
        "long lived tree of depth {}\t check: {}",
        max_depth,
        long_lasting_tree.check_tree()
    );

    println!("time: {}ms", start.elapsed().as_millis());
}

fn main() {
    bench_parallel();
}